### ISSUE TITLE
often recognize 'u' wrongly

### ISSUE THREAD

**Issue by kojit** [weinman/cnn_lstm_ctc_ocr#42]
Hello,

I trained your model with mjsynth dataset and default parameter settings over 1000000 steps.
I found that the model often miss-recognizes character 'u'.
Do you have any thoughts about what the cause might be?

**Comment by kojit** [comment]
I've done training the model 2**21 steps.
But it still cannot recognize 'u' correctly, I also found that it doesn't recognize 'q' at all.
I tested with several images and found that the probability of 'u' and 'q' before the CTC layer are always 0.
Does anyone have such experiences?

**Comment by kojit** [comment]
It's wired though, after I only changed the CNN model to Shi et al.'s CRNN architecture version, then it recognizes 'u'.

**Comment by weinman** [comment]
What is the training loss? Validation loss? What values does `test.py` report on the test data? 

The default training parameters sometimes can get stuck (quite early in training) in a poor local minimum. I've never investigated specific character-level confusions/probabilities, but I definitely don't see this behavior in my own experiences.

To avoid local minima, I have set up an alternative training schedule that starts with a small batch size and increases  from 16 to 128 as the step size (no staircase) decreases from .0001 down to .000003. (See [Takase et al.](https://www.mitpressjournals.org/doi/abs/10.1162/neco_a_01089))

**Comment by kojit** [comment]
Thanks for your reply.

test.py shows as follows although I didn't use the entire test set because it's too slow.

 {'total_num_labels': 144942, 'total_num_sequence_errs': 3892, 'total_num_label_errors': 6711, 'mean_label_error': 0.04630127913234259, 'loss': 1.5078024, 'total_num_sequences': 17837, 'mean_sequence_error': 0.21819812748780623, 'global_step': 2097152}

I understand that you've never seen this problem and you think it's a local minimum.
I'd like to try changing battch size.

**Comment by weinman** [comment]
Those label error rates and sequence error rates seem pretty reasonable. Maybe its not a local minimum.

That loss seems a bit high, but I just realized that `test.py` probably reports only the last test batch's loss (rather than a cumulative average, which it should). 

What's the smoothed training loss (i.e., as reported in tensorboard)? (Say with a smoothing factor of something like 0.95.)

My training schedule is as follows:

| Batch Size |  Learning Rate  |  Steps (Cumulative) 
|-----|------|-------
| 16    | 1e-4 |   2^16
| 32    | 3e-5 |   2^18
| 64    | 3e-5 |  2^19
| 128 |  1e-5 | 2^19 + 2^18
| 128 | 3e-6 | 2^20

**Comment by kojit** [comment]
Smoothed training loss is 1.072.

**Comment by weinman** [comment]
Oh yeah that's probably not very good. You want it down around 0.4–0.5. 

The colors below indicate the training sessions in the table above.

<img width="744" alt="image" src="https://user-images.githubusercontent.com/3750004/46776356-d2668c80-ccd0-11e8-9f32-436ceccf3806.png">

**Comment by kojit** [comment]
I've done training the model 2**21 steps.
What was wrong...?

**Comment by kojit** [comment]
Thanks. I will try with that and report later.

**Comment by sahilbandar** [comment]
Same here, I've trained this for the 1^21 epochs, It is not able to recognise 8 and 9.

**Comment by weinman** [comment]
@sahilbandar Just decay rate=1, as well as the batch size, learning rate, max number of steps, (and tune from) to set the schedule [noted above](https://github.com/weinman/cnn_lstm_ctc_ocr/issues/42#issuecomment-428793851).
